community/kernel/src/main/java/org/neo4j/kernel/api/ExecutingQuery.java

```diff
@@ -20,41 +20,64 @@
 package org.neo4j.kernel.api;
 
 import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLongFieldUpdater;
 
+import org.apache.commons.lang3.builder.ToStringBuilder;
+
+import org.neo4j.kernel.impl.locking.Locks;
 import org.neo4j.kernel.impl.query.QuerySource;
+import org.neo4j.storageengine.api.lock.ResourceType;
+import org.neo4j.time.CpuClock;
+import org.neo4j.time.SystemNanoClock;
 
-import static java.lang.String.format;
+import static java.util.concurrent.atomic.AtomicLongFieldUpdater.newUpdater;
 
 /**
  * Represents a currently running query.
  */
 public class ExecutingQuery
 {
+    private static final AtomicLongFieldUpdater<ExecutingQuery> WAIT_TIME =
+            newUpdater( ExecutingQuery.class, "waitTimeNanos" );
     private final long queryId;
+    private final Locks.Tracer lockTracer = ExecutingQuery.this::waitForLock;
```

Comment:
I do not fully agree.

I agree that it should be possible to have a tracer provided through configuration. But I don't think that excludes being able to use the event based tracer pattern for other purposes as well. The important part of that is just that if one is needed for some feature, it shouldn't disable other features. This is one reason why we included the `CombinedTracer` so that the configured system-tracer can be combined with the query specific one(s).